crates/rover-client/src/query/partial/fetch.rs

```diff
@@ -0,0 +1,187 @@
+use crate::blocking::StudioClient;
+use crate::RoverClientError;
+use graphql_client::*;
+
+#[derive(GraphQLQuery)]
+// The paths are relative to the directory where your `Cargo.toml` is located.
+// Both json and the GraphQL schema language are supported as sources for the schema
+#[graphql(
+    query_path = "src/query/partial/fetch.graphql",
+    schema_path = "schema.graphql",
+    response_derives = "PartialEq, Debug, Serialize, Deserialize",
+    deprecated = "warn"
+)]
+/// This struct is used to generate the module containing `Variables` and
+/// `ResponseData` structs.
+/// Snake case of this name is the mod name. i.e. fetch_subgraph_query
+pub struct FetchSubgraphQuery;
+
+/// The main function to be used from this module. This function fetches a
+/// schema from apollo studio and returns it in either sdl (default) or json format
+pub fn run(
+    variables: fetch_subgraph_query::Variables,
+    client: &StudioClient,
+    // we can't specify this as a variable in the op, so we have to filter the
+    // operation response by this name
+    service_name: &str,
+) -> Result<String, RoverClientError> {
+    let response_data = client.post::<FetchSubgraphQuery>(variables)?;
+    let services = get_services_from_response_data(response_data)?;
+    get_sdl_for_service(services, service_name)
+    // if we want json, we can parse & serialize it here
+}
+
+type ServiceList = Vec<fetch_subgraph_query::FetchSubgraphQueryServiceImplementingServicesOnFederatedImplementingServicesServices>;
+fn get_services_from_response_data(
+    response_data: fetch_subgraph_query::ResponseData,
+) -> Result<ServiceList, RoverClientError> {
+    let service_data = match response_data.service {
+        Some(data) => Ok(data),
+        None => Err(RoverClientError::NoService),
+    }?;
+
+    // get list of services
+    let services = match service_data.implementing_services {
+        Some(services) => Ok(services),
+        None => Err(RoverClientError::HandleResponse {
+            msg: "There are no implementing services in this graph".to_string(),
+        }),
+    }?;
+
+    match services {
+        fetch_subgraph_query::FetchSubgraphQueryServiceImplementingServices::FederatedImplementingServices (services) => {
+            Ok(services.services)
+        },
+        _ => {
+            unreachable!("This case shouldn't be possible, since the operation doesn't select it!")
```

Comment:
what does this mean?

Reply:
As I was writing my response for this comment, I think I realized this case actually _was_ potentially possible. So I'll change this to a regular error I think. But the idea was that we select `... on FederatedImplementingServices` in the operation, where it's possible to have federated services or regular ones. So we're not actually selecting any fields on non-implementing services. But I think since we're still technically selecting `__typename` on all services, there is a potential path that could lead us to that other branch of the enum

(I realize none of this probably makes sense. I'm not sure how to explain this is text very well haha)